Speed up the maxima _commands() list

orlitzky opened the pull request:
Reimplement the `_commands()` method that queries maxima for a list of valid commands. Gives about a 9x speedup here, and should eliminate one of the CI "slow doctest" warnings.

orlitzky committed:
src/sage/interfaces/maxima_abstract.py: reimplement _commands()

The maxima _commands() method was implemented in terms of the
tab-completion method, completions(), which goes one-letter-at-a-time
and is very slow as a result. We reimplement it directly to call
maxima's apropos() with the empty string. This gets all commands in
one shot. We also clean up the list of commands a little by excluding
some junk results. The diff (eliminated elements) from the old list to
the new one is,

  {'SPLITS\\IN\\Q',
   'erf_%iargs',
   'exp-form',
   'is-boole-eval',
   'is-boole-verify',
   'maybe-boole-eval',
   'maybe-boole-verify',
   'sstatus-aux',
   'time\\/\\/call',
   'unknown\\?',
   'zeta%pi'}

The faster method should help avoid warnings in the CI about the test
being too slow.

orlitzky committed:
src/sage/interfaces/maxima_abstract.py: drop "verbose" from _commands()

This parameter was recently made obsolete, and as this is an internal
method, we can just delete it.

fchapoton commented on src/sage/interfaces/maxima_abstract.py:
```diff
@@ -291,36 +292,59 @@ def completions(self, s, verbose=True):
+        # ' toeplitz', etc.
+        #
+        bad_chars = ( "-", "/", "?", "%" )
+        return [ c.strip()
```
no space inside brackets and parentheses, please, as per pep8

orlitzky replied:
done

orlitzky committed:
src/sage/interfaces/maxima_abstract.py: whitespace (PEP8) in _commands()

orlitzky committed:
src/sage/interfaces/maxima_abstract.py: use maxima_calculus in a doctest

It probably doesn't matter, but we should prefer the maxima library
interface anywhere that performance might be an issue. We now test
the _commands() method with maxima_calculus (a library interface)
rather than "maxima" (a pexpect interface).

orlitzky committed:
src/sage/interfaces/maxima_abstract.py: more bad chars in _commands()

If we use a "\" instead of "-" in the list of bad characters, we catch
all of the existing bad names, but get things like "SPLITS\\ IN\\ Q"
as well.

orlitzky committed:
src/sage/interfaces/maxima_abstract.py: micro-optimize _commands()

The _commands() method checks the first character of each command to
make sure that it starts with a letter. This _was_ searching a tuple
(containing 'a' to 'Z'), but it seems to be much faster if we
concatenate them all into one string and search that instead.

fchapoton commented on src/sage/interfaces/maxima_abstract.py:
```diff
@@ -291,36 +292,63 @@ def completions(self, s, verbose=True):
+        return [c.strip()
+                for c in all_names
+                if c
+                and (c[0] in a_to_Z or c[0] == " ")
```
one first pass to strip would be better

orlitzky replied:
Should be done now, I found a clever way to do it without making a second pass and without resorting to a "for" loop.

orlitzky committed:
src/sage/interfaces/maxima_abstract.py: clarify a comprehension

Using python's walrus operator we can simplify the list comprehension
that builds the list of maxima _commands(). Essentially this allows
us to save the result of foo.strip() and work with that rather than
the original unstripped string foo.

fchapoton reviewed (approved):
good, thanks
